DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs

```diff
@@ -366,6 +368,29 @@ CountInfo CreateCountInfo(AttributeData marshalUsingData, ImmutableHashSet<strin
             return NoCountInfo.Instance;
         }
 
+        private TypePositionInfo? CreateForParamIndex(AttributeData attrData, int paramIndex, ImmutableHashSet<string> inspectedElements)
+        {
+            if (!(contextSymbol is IMethodSymbol method && 0 <= paramIndex && paramIndex < method.Parameters.Length))
+            {
+                return null;
+            }
+            IParameterSymbol param = method.Parameters[paramIndex];
+            try
+            {
+                return TypePositionInfo.CreateForParameter(
+                    param,
+                    ParseMarshallingInfo(param.Type, param.GetAttributes(), inspectedElements.Add(param.Name)), compilation) with
+                { ManagedIndex = paramIndex };
+            }
+            // Specifically catch the exception when we're trying to inspect the element that started the cycle.
+            // This ensures that we've unwound the whole cycle so when we return, there will be no cycles in the count info.
+            catch (CyclicalCountElementInfoException ex) when (ex.StartOfCycle == param.Name)
```

Comment:
So we are using exceptions for control flow here? Not a fan of this.

Reply:
We're using it to capture the exceptional case of cyclical count info, which will always report an error.

Reply:
So the `when` class will always evaluate to the correct frame right? We are using it simply to report at the correct level?

Reply:
Yes, the when clause is to make sure we are reporting at the correct level.